common/solver/batch_gmres_kernels.hpp.inc

```diff
@@ -35,403 +35,210 @@ namespace {
-         tid += blockDim.x) {
-        const int row_index =
-            (tid / V_shared_entry.num_rhs) % x_shared_entry.num_rows;
-
-        const int basis_vec_index =
-            (tid / V_shared_entry.num_rhs) / x_shared_entry.num_rows;
-
-        const int rhs_index = tid % V_shared_entry.num_rhs;
-
-        V_shared_entry
-            .values[(basis_vec_index * x_shared_entry.num_rows + row_index) *
-                        V_shared_entry.stride +
-                    rhs_index] = zero<ValueType>();
+    for (int li = threadIdx.x; li < (restart + 1) * restart; li += blockDim.x) {
+        H_shared_entry[li] = zero<ValueType>();
     }
 
-
-    for (int li = threadIdx.x;
-         li < cs_shared_entry.num_rows * cs_shared_entry.num_rhs;
-         li += blockDim.x) {
-        const int r = li / cs_shared_entry.num_rhs;
-        const int c = li % cs_shared_entry.num_rhs;
-
-        cs_shared_entry.values[r * cs_shared_entry.stride + c] =
-            zero<ValueType>();
+    for (int i = threadIdx.x; i < a_gl_entry.num_rows * (restart + 1);
+         i += blockDim.x) {
+        V_shared_entry[i] = zero<ValueType>();
     }
 
-
-    for (int li = threadIdx.x;
-         li < sn_shared_entry.num_rows * sn_shared_entry.num_rhs;
-         li += blockDim.x) {
-        const int r = li / sn_shared_entry.num_rhs;
-        const int c = li % sn_shared_entry.num_rhs;
-
-        sn_shared_entry.values[r * sn_shared_entry.stride + c] =
-            zero<ValueType>();
+    for (int li = threadIdx.x; li < restart; li += blockDim.x) {
+        cs_shared_entry[li] = zero<ValueType>();
+        sn_shared_entry[li] = zero<ValueType>();
     }
 }
 
 
 template <typename ValueType>
 __device__ __forceinline__ void update_v_naught_and_s(
-    const gko::batch_dense::BatchEntry<const ValueType> &z_shared_entry,
-    const gko::batch_dense::BatchEntry<ValueType> &V_shared_entry,
-    const gko::batch_dense::BatchEntry<ValueType> &s_shared_entry,
-    const gko::batch_dense::BatchEntry<typename gko::remove_complex<ValueType>>
-        &tmp_norms_shared_entry,
-    const uint32 &converged)
+    group::thread_block_tile<config::warp_size> &warp_grp, const int num_rows,
+    const int restart, const ValueType *const z_shared_entry,
+    ValueType *const V_shared_entry, ValueType *const s_shared_entry)
 {
     using real_type = typename gko::remove_complex<ValueType>;
 
-    const gko::batch_dense::BatchEntry<ValueType> v_naught_shared_entry{
-        V_shared_entry.values, V_shared_entry.stride, z_shared_entry.num_rows,
-        V_shared_entry.num_rhs};
-
-
-    const gko::batch_dense::BatchEntry<real_type> &z_norms_shared_entry =
-        tmp_norms_shared_entry;
-
-    compute_norm2(z_shared_entry, z_norms_shared_entry, converged);
-
+    __shared__ real_type tmp_norm;
+    if (threadIdx.x / config::warp_size == 0) {
+        single_compute_norm2(warp_grp, num_rows, z_shared_entry, tmp_norm);
+    }
     __syncthreads();
```

Comment:
I get the idea , but I guess this means that a lot of warps are stalling here ?

Reply:
Yes. According to Isha's tests, this should be faster than trying to use all warps, at least for the matrices we are working with right now.